Ska/Sun/sun.py

```diff
@@ -2,14 +2,29 @@
 """
 Utility for calculating sun position and pitch angle.
 """
+from pathlib import Path
+import numpy as np
+from math import cos, sin, acos, atan2, asin, pi, radians, degrees
+from astropy.table import Table
 
 from Quaternion import Quat
 from chandra_aca.transform import radec_to_eci
 from Chandra.Time import DateTime
-from math import cos, sin, acos, atan2, asin, pi, radians, degrees
-import numpy as np
 import Ska.quatutil
 
+ROLL_TABLE = Table.read(str(Path(__file__).parent / 'data' / 'pitch_rolldev.csv'),
+                        format='ascii.basic',
+                        guess=False, delimiter=',')
+
+
+def allowed_rolldev(pitch):
+    """Get allowed roll deviation (off-nominal roll) for the given ``pitch``.
+    :param pitch: Sun pitch angle (deg)
+    :returns: Roll deviation (deg)
+    """
+    idx = np.searchsorted(ROLL_TABLE['pitch'], pitch, side='right')
```

Comment:
I am not sure this is correct for the new table, which is no longer done as a stairstep but rather samples values. I'd say that `np.interp()` is more appropriate. This should (by default) give the right behavior for pitch values outside the table range.

Reply:
"I am not sure this is correct for the new table, which is no longer done as a stairstep but rather samples values." The table we were using was the 278 line table (not the 62 line matlab constraints stairstep..)  but I still thought it was defined as a stairstep.  Maybe we should check?  Is the right behavior for pitch values outside the table range 0 or error?

Reply:
After that long slack conversation I think the approach below could work.  It is close but always conservative by choosing the minimum of points to the left and to the right:
```
idx0 = np.searchsorted(ROLL_TABLE['pitch'], pitch, side='left')
idx1 = np.searchsorted(ROLL_TABLE['pitch'], pitch, side='right')
# These need some bound checking at the edges
val0 = ROLL_TABLE['rolldev'][idx0]
val1 = ROLL_TABLE['rolldev'][idx1]
out = np.minimum(val0, val1)  # works even for a vector input for `pitch`
return out
```

Reply:
>  Is the right behavior for pitch values outside the table range 0 or error?

The right answer is 0.0.

Reply:
I don't quite understand the bounds checking here, but how do you want to handle the >= 180 cases then?
I'd probably just stick some logic to define outside of bounds as 0 but...


```
In [6]: allowed_rolldev(180)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-6-c0f6f350c041> in <module>
----> 1 allowed_rolldev(180)

~/git/Ska.Sun/Ska/Sun/sun.py in allowed_rolldev(pitch)
     30     idx1 = np.searchsorted(ROLL_TABLE.val['pitch'], pitch, side='right')
     31     # These need some bound checking at the edges
---> 32     val0 = ROLL_TABLE.val['rolldev'][idx0]
     33     val1 = ROLL_TABLE.val['rolldev'][idx1]
     34     out = np.minimum(val0, val1)  # works even for a vector input for `pitch`

astropy/table/_column_mixins.pyx in astropy.table._column_mixins._ColumnGetitemShim.__getitem__()

astropy/table/_column_mixins.pyx in astropy.table._column_mixins.base_getitem()

astropy/table/_column_mixins.pyx in astropy.table._column_mixins.column_getitem()

IndexError: index 13340 is out of bounds for axis 0 with size 13340
```

Reply:
And did you want to support vector pitch or not?

Reply:
Yes on vector pitch.

Reply:
This might work:
```
idx_max =  len(ROLL_TABLE.val) - 1
idx0 = np.clip(idx0, 0, idx_max)
idx1 = np.clip(idx1, 0, idx_max)
```

Reply:
If you meant the function as a whole would look like 
```
    idx0 = np.searchsorted(ROLL_TABLE.val['pitch'], pitch, side='left')
    idx1 = np.searchsorted(ROLL_TABLE.val['pitch'], pitch, side='right')
    idx_max =  len(ROLL_TABLE.val) - 1
    idx0 = np.clip(idx0, 0, idx_max)
    idx1 = np.clip(idx1, 0, idx_max)
    # These need some bound checking at the edges
    val0 = ROLL_TABLE.val['rolldev'][idx0]
    val1 = ROLL_TABLE.val['rolldev'][idx1]
    return np.minimum(val0, val1) 
```
It looks like for pitch >= 180 idx0 and idx1 both end up set at the last element so they stick at 18.748.  I was thinking about something else, but if you want to play with this that's fine.

Reply:
If you meant this in combination with adding another row to the data file or the table with pitch(180) = 0, that works.

Reply:
Right, it looks like it mostly worked, but side='left' really doesn't give you the index left of the the value you want... it still gives you the index that the value would have.  Subtracting 1 from the side='right' answer to get idx0 might be a little closer?  I've updated for that for now.

Reply:
Good catch with the testing. I had fallen into the trap of thinking about the "left" and "right" values incorrectly, even though I should know better. Bah. The solution of subtracting 1 is correct!